### ISSUE TITLE
Number/hash sign (#) is escaped

### ISSUE THREAD

**Issue by fominok** [kivikakk/comrak#509]
Hey, I'm using comrak to build an app on top of markdown-oxide, that uses tags `#like` `#this`;

Tweaking a source AST and writing it back to original files works great, however, tags are unnecessarily rewritten as `\#this`.

If it's about distinguishing from headers, those require a whitespace between sign and title, so this escape seems redundant, other than that I don't know why it could be needed

**Comment by kivikakk** [comment]
Hi! This is part of the CommonMark escaping strategy we inherited from our upstream. There's some truly ancient code in there; just now I `git blame`d something I didn't recognise even slightly and it was me, almost 8 years ago.

The extant strategy is that, in regular text, we simply always escape a bunch of characters, hash `#` included:

https://github.com/kivikakk/comrak/blob/8d1e90c5c125011bbc7d99332de92dc3cfe9489f/src/cm.rs#L202-L222

We obviously can be smarter about this, but it requires a bit more intelligence than the formatter currently provides for (viz. almost none — right now the function outputting a character has no context whatsoever).

Would you be willing to try submitting a PR to fix this? An acceptable minimal improvement would, for instance, pass through `buf` and `i` (instead of `buf[i]`) to `self.outc` on line 170, and then — in `outc` — not escape `#` if there's a following character which isn't whitespace. This gets you over the line without us rewriting the formatter.

**Comment by fominok** [comment]
Hey, yes that was my initial idea here, to match on string there instead of a char and the string would be `"# "`

Thanks, I'll make a PR today

**Comment by fominok** [comment]
I'm afraid it goes deeper, even exclamation marks are escaped

**Comment by kivikakk** [comment]
Yes, if you look at the code snippet I've quoted above, quite a lot of things are escaped — anything that could form part of valid markup. (For exclamation mark `!`, it's images: `![blah](blah)`.)

**Comment by charlottia** [comment]
Hiya! I've tried putting together an option for this at https://github.com/kivikakk/comrak/pull/523 — @fominok, if you're still interested, could you give that branch a try and let me know if it works OK?

**Comment by fominok** [comment]
hey @charlottia and thanks for your effort

I've found myself expecting more and more from the parser (and renderer back to MD document) that even doesn't fit into the spec anymore, so I'm using a fork now that doesn't escape at all among other things needed for my project

**Comment by charlottia** [comment]
No problems, thanks for letting us know! :D 

Will leave this issue open until I get some feedback on that PR and see if it helps this case for others.
